drivers/ft5x06/ft5x06.c

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (C) 2021 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_ft5x06
+ * @{
+ *
+ * @file
+ * @brief       Device driver implementation for the FT5x06 touch driver
+ *
+ * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
+ *
+ * @}
+ */
+
+#include <errno.h>
+#include <stdint.h>
+
+#include "periph/i2c.h"
+#include "periph/gpio.h"
+
+#include "ztimer.h"
+
+#include "ft5x06.h"
+#include "ft5x06_constants.h"
+#include "ft5x06_params.h"
+
+#define ENABLE_DEBUG    0
+#include "debug.h"
+
+#define FT5X06_BUS              (dev->params.i2c)
+#define FT5X06_ADDR             (dev->params.addr)
+
+#define FT5X06_RESET_DELAY_MS   (200)
+
+static void _gpio_irq(void *arg)
+{
+    const ft5x06_t *dev = (const ft5x06_t *)arg;
+
+    assert(dev);
+
+    if (dev->cb) {
+        dev->cb(dev->cb_arg);
+    }
+}
+
+int ft5x06_init(ft5x06_t *dev, const ft5x06_params_t *params, ft5x06_event_cb_t cb, void *arg)
+{
+    dev->params = *params;
+    dev->cb = cb;
+    dev->cb_arg = arg;
+
+    /* Wait at least 200ms after power up before accessing registers */
+    ztimer_sleep(ZTIMER_MSEC, FT5X06_RESET_DELAY_MS);
+
+    i2c_acquire(FT5X06_BUS);
+
+    uint8_t vendor_id = 0;
+    if (i2c_read_reg(FT5X06_BUS, FT5X06_ADDR, FT5X06_G_VENDOR_ID_REG, &vendor_id, 0) != 0) {
+        i2c_release(FT5X06_BUS);
+        return -EPROTO;
+    }
+
+    if (vendor_id != FT5X06_VENDOR_ID) {
+        DEBUG("[ft5x06] init: invalid vendor ID: '0x%02x' (expected: 0x%02x)\n",
+                vendor_id, FT5X06_VENDOR_ID);
+        i2c_release(FT5X06_BUS);
+        return -ENODEV;
+    }
+
+    /* Auto-calibrate if needed */
+    if (IS_ACTIVE(FT5X06_AUTO_CALIB_NEEDED)) {
+        DEBUG("[ft5x06] init: enable device auto-calibration\n");
+        i2c_write_reg(FT5X06_BUS, FT5X06_ADDR, FT5X06_G_AUTO_CLB_MODE_REG, 0, 0);
+    }
+
+    /* Configure interrupt */
+    if (gpio_is_valid(dev->params.int_pin)) {
+        DEBUG("[ft5x06] init: configuring touchscreen interrupt\n");
+        gpio_init_int(dev->params.int_pin, GPIO_IN, GPIO_RISING, _gpio_irq, dev);
```

Comment:
Why not
```suggestion
        gpio_init_int(dev->params.int_pin, GPIO_IN, GPIO_RISING, cb, arg);
```

Reply:
The cb attribute can be changed at runtime, that's part of the touch_dev API (set_event_callback). I think that's the reason IIRC.

Reply:
hm I see, I guess #13925 would avoid the indirection.

But then again, since this is not time critical, why not just call `gpio_init_int()` again?

Reply:
I think it's done like that for the stmpe811, I just followed the same logic. Maybe we can improve later in both place (and also for the cst816s that is doing something similar).

Reply:
Just because some other driver does something unnecessary doesn't mean we have to copy it 😉 
Otherwise people will get the idea that this is the pattern one should follow.

Reply:
Ok, I agree that your proposal is better. I tried locally and it saves some ROM/RAM. I'll change that and verify it still works.

Reply:
I tested both `tests/driver_ft5x06` and `tests/touch_dev` with the direct cb/arg mapping and they work the same, the firmware is just a little lighter :+1: